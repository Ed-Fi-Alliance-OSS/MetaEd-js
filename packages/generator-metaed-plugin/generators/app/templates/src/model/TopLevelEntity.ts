import { MetaEdEnvironment, EnhancerResult, TopLevelEntity, getAllTopLevelEntitiesForNamespaces } from 'metaed-core';

// Model that defines a new data element that hangs off of a TopLevelEntity
// File named the same as the parent object type
// Exactly one parent object type per file

// Definition of an additional data element for all top level entities
export type TopLevelEntityEdfiOdsApi = {
  exampleName: string;
};

// Model object initialization is a "setup" enhancer named prefixed as the parent object type
const enhancerName = 'TopLevelEntitySetupEnhancer';

// exampleName is defined as not null, so intitalize TopLevelEntity with an empty string,
// creating the plugin-specfic slot "orgExample" (the plugin shortname) on the "data" sub-object, if not already there
export function addTopLevelEntityOrgExampleTo(topLevelEntity: TopLevelEntity) {
  if (topLevelEntity.data.orgExample == null) topLevelEntity.data.orgExample = {};

  Object.assign(topLevelEntity.data.orgExample, {
    exampleName: '',
  });
}

// standard enhancer function signature
export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
  // Loop over all TopLevelEntity objects in all namespaces
  getAllTopLevelEntitiesForNamespaces([...metaEd.namespace.values()]).forEach(entity => {
    // Initialize each entity
    addTopLevelEntityOrgExampleTo(entity);
  });

  return {
    enhancerName,
    success: true,
  };
}
