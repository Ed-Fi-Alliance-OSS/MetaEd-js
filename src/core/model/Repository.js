// @flow
import type { Association } from './Association';
import type { AssociationExtension } from './AssociationExtension';
import type { AssociationSubclass } from './AssociationSubclass';
import type { Choice } from './Choice';
import type { Common } from './Common';
import type { CommonExtension } from './CommonExtension';
import type { Descriptor } from './Descriptor';
import type { DomainEntity } from './DomainEntity';
import type { DomainEntityExtension } from './DomainEntityExtension';
import type { DomainEntitySubclass } from './DomainEntitySubclass';
import type { Enumeration } from './Enumeration';
import type { Interchange } from './Interchange';
import type { InterchangeExtension } from './InterchangeExtension';
import type { MapTypeEnumeration } from './MapTypeEnumeration';
import type { NamespaceInfo } from './NamespaceInfo';
import type { SchoolYearEnumeration } from './SchoolYearEnumeration';
import type { SharedDecimal } from './SharedDecimal';
import type { SharedInteger } from './SharedInteger';
import type { SharedString } from './SharedString';
import type { Domain } from './Domain';
import type { Subdomain } from './Subdomain';
import type { PropertyRepository } from './property/PropertyRepository';
import { propertyRepositoryFactory } from './property/PropertyRepository';

export class EntityRepository {
  association: Map<string, Association>;
  associationExtension: Map<string, AssociationExtension>;
  associationSubclass: Map<string, AssociationSubclass>;
  choice: Map<string, Choice>;
  common: Map<string, Common>;
  commonExtension: Map<string, CommonExtension>;
  descriptor: Map<string, Descriptor>;
  domain: Map<string, Domain>;
  domainEntity: Map<string, DomainEntity>;
  domainEntityExtension: Map<string, DomainEntityExtension>;
  domainEntitySubclass: Map<string, DomainEntitySubclass>;
  enumeration: Map<string, Enumeration>;
  interchange: Map<string, Interchange>;
  interchangeExtension: Map<string, InterchangeExtension>;
  mapTypeEnumeration: Map<string, MapTypeEnumeration>;
  namespaceInfo: Array<NamespaceInfo>;
  schoolYearEnumeration: Map<string, SchoolYearEnumeration>;
  sharedDecimal: Map<string, SharedDecimal>;
  sharedInteger: Map<string, SharedInteger>;
  sharedString: Map<string, SharedString>;
  subdomain: Map<string, Subdomain>;
}

export function entityRepositoryFactory(): EntityRepository {
  return Object.assign(new EntityRepository(), {
    association: new Map(),
    associationExtension: new Map(),
    associationSubclass: new Map(),
    choice: new Map(),
    common: new Map(),
    commonExtension: new Map(),
    descriptor: new Map(),
    domain: new Map(),
    domainEntity: new Map(),
    domainEntityExtension: new Map(),
    domainEntitySubclass: new Map(),
    enumeration: new Map(),
    interchange: new Map(),
    interchangeExtension: new Map(),
    mapTypeEnumeration: new Map(),
    namespaceInfo: [],
    schoolYearEnumeration: new Map(),
    sharedDecimal: new Map(),
    sharedInteger: new Map(),
    sharedString: new Map(),
    subdomain: new Map(),
  });
}

export class Repository {
  configuration: any;
  entity: EntityRepository;
  property: PropertyRepository;
}

export function repositoryFactory(): Repository {
  return Object.assign(new Repository(), {
    configuration: null,
    entity: entityRepositoryFactory(),
    property: propertyRepositoryFactory(),
  });
}

export type MostEntities =
  Association |
  AssociationSubclass |
  Choice |
  Common |
  DomainEntity |
  DomainEntitySubclass |
  SharedDecimal |
  SharedInteger |
  SharedString;

// Domains, Subdomains, Interchanges, Enumerations and Descriptors don't have standard cross entity naming issues
// and extension entities don't define a new identifier
export function entitiesNeedingDuplicateChecking(repository: Repository): Array<MostEntities> {
  const result: Array<MostEntities> = [];
  result.push(...repository.entity.association.values());
  result.push(...repository.entity.associationSubclass.values());
  result.push(...repository.entity.choice.values());
  result.push(...repository.entity.common.values());
  result.push(...repository.entity.domainEntity.values());
  result.push(...repository.entity.domainEntitySubclass.values());
  result.push(...repository.entity.sharedDecimal.values());
  result.push(...repository.entity.sharedInteger.values());
  result.push(...repository.entity.sharedString.values());
  return result;
}
