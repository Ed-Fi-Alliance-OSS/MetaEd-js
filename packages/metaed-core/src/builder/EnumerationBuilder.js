// @flow
import type { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newEnumeration, asEnumeration } from '../model/Enumeration';
import { newSchoolYearEnumeration } from '../model/SchoolYearEnumeration';
import type { EnumerationSourceMap } from '../model/Enumeration';
import type { EnumerationItem } from '../model/EnumerationItem';
import { newEnumerationItem, NoEnumerationItem } from '../model/EnumerationItem';
import { extractDocumentation, extractShortDescription, squareBracketRemoval, isErrorText } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import type { MetaEdEnvironment } from '../MetaEdEnvironment';
import type { ValidationFailure } from '../validator/ValidationFailure';
import { sourceMapFrom } from '../model/SourceMap';

export class EnumerationBuilder extends TopLevelEntityBuilder {
  currentEnumerationItem: EnumerationItem;

  constructor(metaEd: MetaEdEnvironment, validationFailures: Array<ValidationFailure>) {
    super(metaEd, validationFailures);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.enteringEntity(newEnumeration);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
        namespaceInfo: this.currentTopLevelEntity.namespaceInfo.sourceMap.type,
      });
    }
  }

  // eslint-disable-next-line no-unused-vars
  exitEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.exitingEntity();
  }

  enterEnumerationName(context: MetaEdGrammar.EnumerationNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    const enumerationName = context.ID().getText();

    // need to differentiate SchoolYear from other enumerations - overwrite with new type
    if (enumerationName === 'SchoolYear') {
      this.enteringEntity(newSchoolYearEnumeration);
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
        namespaceInfo: this.currentTopLevelEntity.namespaceInfo.sourceMap.type,
      });
    }

    this.enteringName(enumerationName);
    Object.assign(this.currentTopLevelEntity.sourceMap, {
      metaEdName: sourceMapFrom(context),
    });
  }

  enterEnumerationItemDocumentation(context: MetaEdGrammar.EnumerationItemDocumentationContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.documentation = extractDocumentation(context);
    this.currentEnumerationItem.sourceMap.documentation = sourceMapFrom(context);
  }

  enterEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    this.currentEnumerationItem = newEnumerationItem();
    this.currentEnumerationItem.namespaceInfo.sourceMap = this.currentTopLevelEntity.namespaceInfo.sourceMap;
    this.currentEnumerationItem.sourceMap.type = sourceMapFrom(context);
  }

  exitEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity || this.currentEnumerationItem === NoEnumerationItem) return;
    asEnumeration(this.currentTopLevelEntity).enumerationItems.push(this.currentEnumerationItem);
    ((this.currentTopLevelEntity.sourceMap: any): EnumerationSourceMap).enumerationItems.push(sourceMapFrom(context));
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterMetaEdId(context: MetaEdGrammar.MetaEdIdContext) {
    if (context.exception || context.METAED_ID() == null || context.METAED_ID().exception != null || isErrorText(context.METAED_ID().getText())) return;
    if (this.currentEnumerationItem !== NoEnumerationItem) {
      this.currentEnumerationItem.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
      this.currentEnumerationItem.sourceMap.metaEdId = sourceMapFrom(context);
    } else {
      super.enterMetaEdId(context);
    }
  }

  enterShortDescription(context: MetaEdGrammar.ShortDescriptionContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.shortDescription = extractShortDescription(context);
    this.currentEnumerationItem.sourceMap.shortDescription = sourceMapFrom(context);
  }
}
