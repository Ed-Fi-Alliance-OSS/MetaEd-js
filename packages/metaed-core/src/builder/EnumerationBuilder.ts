import { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newEnumeration, asEnumeration } from '../model/Enumeration';
import { newSchoolYearEnumeration } from '../model/SchoolYearEnumeration';
import { EnumerationSourceMap } from '../model/Enumeration';
import { EnumerationItem } from '../model/EnumerationItem';
import { newEnumerationItem, NoEnumerationItem } from '../model/EnumerationItem';
import { extractDocumentation, extractShortDescription, isErrorText } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { MetaEdEnvironment } from '../MetaEdEnvironment';
import { ValidationFailure } from '../validator/ValidationFailure';
import { sourceMapFrom } from '../model/SourceMap';

/**
 * An ANTLR4 listener that creates Enumeration entities.
 */
export class EnumerationBuilder extends TopLevelEntityBuilder {
  currentEnumerationItem: EnumerationItem;

  constructor(metaEd: MetaEdEnvironment, validationFailures: ValidationFailure[]) {
    super(metaEd, validationFailures);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.enteringEntity(newEnumeration);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
      });
    }
  }

  exitEnumeration(_context: MetaEdGrammar.EnumerationContext) {
    this.exitingEntity();
  }

  enterEnumerationName(context: MetaEdGrammar.EnumerationNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    const enumerationName = context.ID().getText();

    // need to differentiate SchoolYear from other enumerations - overwrite with new type
    if (enumerationName === 'SchoolYear') {
      this.enteringEntity(newSchoolYearEnumeration);
      Object.assign(this.currentTopLevelEntity.sourceMap, {
        type: sourceMapFrom(context),
      });
    }

    this.enteringName(enumerationName);
    Object.assign(this.currentTopLevelEntity.sourceMap, {
      metaEdName: sourceMapFrom(context),
    });
  }

  enterEnumerationItemDocumentation(context: MetaEdGrammar.EnumerationItemDocumentationContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.documentation = extractDocumentation(context);
    this.currentEnumerationItem.sourceMap.documentation = sourceMapFrom(context);
  }

  enterEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    this.currentEnumerationItem = newEnumerationItem();
    this.currentEnumerationItem.sourceMap.type = sourceMapFrom(context);
  }

  exitEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity || this.currentEnumerationItem === NoEnumerationItem) return;
    asEnumeration(this.currentTopLevelEntity).enumerationItems.push(this.currentEnumerationItem);
    (this.currentTopLevelEntity.sourceMap as EnumerationSourceMap).enumerationItems.push(sourceMapFrom(context));
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterShortDescription(context: MetaEdGrammar.ShortDescriptionContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.shortDescription = extractShortDescription(context);
    this.currentEnumerationItem.sourceMap.shortDescription = sourceMapFrom(context);
  }
}
