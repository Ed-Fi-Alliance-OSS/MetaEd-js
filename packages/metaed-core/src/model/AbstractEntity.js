// @flow
import deepFreeze from 'deep-freeze';
import { TopLevelEntity, TopLevelEntitySourceMap, newTopLevelEntity } from './TopLevelEntity';
import type { SourceMap } from './SourceMap';
import type { ModelBase } from './ModelBase';

export class AbstractEntitySourceMap extends TopLevelEntitySourceMap {
  isAbstract: ?SourceMap;
}

export class AbstractEntity extends TopLevelEntity {
  isAbstract: boolean;
  sourceMap: TopLevelEntitySourceMap | AbstractEntitySourceMap;
}

export function newAbstractEntity(): AbstractEntity {
  return Object.assign(new AbstractEntity(), newTopLevelEntity(), {
    type: 'domainEntity',
    typeHumanizedName: 'Abstract Entity',
    isAbstract: true,
    sourceMap: new AbstractEntitySourceMap(),
  });
}

export const NoAbstractEntity: AbstractEntity = deepFreeze(
  Object.assign(newAbstractEntity(), {
    metaEdName: 'NoAbstractEntity',
  }),
);

export const asAbstractEntity = (x: ModelBase): AbstractEntity => ((x: any): AbstractEntity);
