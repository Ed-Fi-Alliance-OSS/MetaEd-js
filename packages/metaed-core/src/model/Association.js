// @flow
import deepFreeze from 'deep-freeze';
import type { TopLevelEntity, TopLevelEntitySourceMap } from './TopLevelEntity';
import { newTopLevelEntity, newTopLevelEntitySourceMap } from './TopLevelEntity';
import type { ModelBase } from './ModelBase';
import type { SourceMap } from './SourceMap';
import { NoSourceMap } from './SourceMap';

export type AssociationSourceMap = {
  isAbstract: SourceMap,
  ...$Exact<TopLevelEntitySourceMap>,
};

export function newAssociationSourceMap(): AssociationSourceMap {
  return {
    ...newTopLevelEntitySourceMap(),
    isAbstract: NoSourceMap,
  };
}

export type Association = {
  sourceMap: AssociationSourceMap,
  ...$Exact<TopLevelEntity>,
  isAbstract: boolean,
};

export function newAssociation(): Association {
  return {
    ...newTopLevelEntity(),
    type: 'association',
    typeHumanizedName: 'Association',
    isAbstract: false,
    sourceMap: newAssociationSourceMap(),
  };
}

export const asAssociation = (x: ModelBase): Association => ((x: any): Association);

export const NoAssociation: Association = deepFreeze({
  ...newAssociation(),
  metaEdName: 'NoAssociation',
});
