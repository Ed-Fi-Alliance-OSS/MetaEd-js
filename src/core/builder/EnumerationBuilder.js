// @flow
import { MetaEdGrammar } from '../../grammar/gen/MetaEdGrammar';
import TopLevelEntityBuilder from './TopLevelEntityBuilder';
import { enumerationFactory } from '../model/Enumeration';
import { schoolYearEnumerationFactory } from '../model/SchoolYearEnumeration';
import type { Enumeration } from '../model/Enumeration';
import type { EnumerationItem } from '../model/EnumerationItem';
import { enumerationItemFactory, NoEnumerationItem } from '../model/EnumerationItem';
import { extractDocumentation, extractShortDescription, squareBracketRemoval } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { EntityRepository } from '../model/Repository';

export default class EnumerationBuilder extends TopLevelEntityBuilder {
  currentEnumerationItem: EnumerationItem;

  constructor(repository: EntityRepository) {
    super(repository);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  // eslint-disable-next-line no-unused-vars
  enterEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.enteringEntity(enumerationFactory);
  }

  // eslint-disable-next-line no-unused-vars
  exitEnumeration(context: MetaEdGrammar.EnumerationContext) {
    this.exitingEntity();
  }

  enterEnumerationName(context: MetaEdGrammar.EnumerationNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception) return;
    const enumerationName = context.ID().getText();

    // need to differentiate SchoolYear from other enumerations - overwrite with new type
    if (enumerationName === 'SchoolYear') {
      this.enteringEntity(schoolYearEnumerationFactory);
    }

    this.enteringName(enumerationName);
  }

  enterEnumerationItemDocumentation(context: MetaEdGrammar.EnumerationItemDocumentationContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.documentation = extractDocumentation(context);
  }

  // eslint-disable-next-line no-unused-vars
  enterEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    this.currentEnumerationItem = enumerationItemFactory();
  }

  // eslint-disable-next-line no-unused-vars
  exitEnumerationItem(context: MetaEdGrammar.EnumerationItemContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity || this.currentEnumerationItem === NoEnumerationItem) return;
    ((this.currentTopLevelEntity: any): Enumeration).enumerationItems.push(this.currentEnumerationItem);
    this.currentEnumerationItem = NoEnumerationItem;
  }

  enterMetaEdId(context: MetaEdGrammar.MetaEdIdContext) {
    if (context.METAED_ID() == null || context.METAED_ID().exception != null) return;
    if (this.currentEnumerationItem !== NoEnumerationItem) {
      this.currentEnumerationItem.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
    } else {
      super.enterMetaEdId(context);
    }
  }

  enterShortDescription(context: MetaEdGrammar.ShortDescriptionContext) {
    if (this.currentEnumerationItem === NoEnumerationItem) return;
    this.currentEnumerationItem.shortDescription = extractShortDescription(context);
  }
}
