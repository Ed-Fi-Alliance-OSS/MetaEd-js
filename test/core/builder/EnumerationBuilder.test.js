// @noflow
import EnumerationBuilder from '../../../src/core/builder/EnumerationBuilder';
import MetaEdTextBuilder from '../MetaEdTextBuilder';
import { repositoryFactory } from '../../../src/core/model/Repository';
import type { Repository } from '../../../src/core/model/Repository';

describe('when building single enumeration', () => {
  const repository: Repository = repositoryFactory();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const metaEdId: string = '1';
  const documentation: string = 'Doc';
  const itemShortDescription: string = 'ItemShortDescription';
  const itemDocumentation: string = 'ItemDocumentation';
  const itemMetaEdId: string = '2';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName, metaEdId)
      .withDocumentation(documentation)
      .withEnumerationItem(itemShortDescription, itemDocumentation, itemMetaEdId)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should build one enumeration', () => {
    expect(repository.enumeration.size).toBe(1);
  });

  it('should be found in repository', () => {
    expect(repository.enumeration.get(entityName)).toBeDefined();
    expect(repository.enumeration.get(entityName).metaEdName).toBe(entityName);
  });

  it('should have correct namespace', () => {
    expect(repository.enumeration.get(entityName).namespaceInfo.namespace).toBe(namespace);
  });

  it('should have correct metaEdId', () => {
    expect(repository.enumeration.get(entityName).metaEdId).toBe(metaEdId);
  });

  it('should have correct project extension', () => {
    expect(repository.enumeration.get(entityName).namespaceInfo.projectExtension).toBe(projectExtension);
  });

  it('should have correct documentation', () => {
    expect(repository.enumeration.get(entityName).documentation).toBe(documentation);
  });

  it('should have no properties', () => {
    expect(repository.enumeration.get(entityName).properties).toHaveLength(0);
  });

  it('should have one enumeration item', () => {
    expect(repository.enumeration.get(entityName).enumerationItems).toHaveLength(1);
  });

  it('should have enumeration item with correct short description', () => {
    expect(repository.enumeration.get(entityName).enumerationItems[0].shortDescription).toBe(itemShortDescription);
  });

  it('should have enumeration item with correct documentation', () => {
    expect(repository.enumeration.get(entityName).enumerationItems[0].documentation).toBe(itemDocumentation);
  });

  it('should have enumeration item with correct metaEdId ', () => {
    expect(repository.enumeration.get(entityName).enumerationItems[0].metaEdId).toBe(itemMetaEdId);
  });
});

describe('when building enumeration without item documentation', () => {
  const repository: Repository = repositoryFactory();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const documentation: string = 'Doc';
  const itemShortDescription: string = 'ItemShortDescription';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName)
      .withDocumentation(documentation)
      .withEnumerationItem(itemShortDescription)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should have correct documentation', () => {
    expect(repository.enumeration.get(entityName).documentation).toBe(documentation);
  });

  it('should have enumeration item with correct short description', () => {
    expect(repository.enumeration.get(entityName).enumerationItems[0].shortDescription).toBe(itemShortDescription);
  });

  it('should have enumeration item with no documentation', () => {
    expect(repository.enumeration.get(entityName).enumerationItems[0].documentation).toBe('');
  });
});

describe('when building multiple enumerations', () => {
  const repository: Repository = repositoryFactory();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName1: string = 'EntityName1';
  const metaEdId1: string = '1';
  const entityName2: string = 'EntityName2';
  const metaEdId2: string = '2';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName1, metaEdId1)
      .withDocumentation('doc')
      .withEnumerationItem('sd1')
      .withEndEnumeration()

      .withStartEnumeration(entityName2, metaEdId2)
      .withDocumentation('doc')
      .withEnumerationItem('sd2')
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should build two enumerations', () => {
    expect(repository.enumeration.size).toBe(2);
  });

  it('should both be found in repository', () => {
    expect(repository.enumeration.get(entityName1)).toBeDefined();
    expect(repository.enumeration.get(entityName1).metaEdName).toBe(entityName1);

    expect(repository.enumeration.get(entityName2)).toBeDefined();
    expect(repository.enumeration.get(entityName2).metaEdName).toBe(entityName2);
  });

  it('should both have correct metaEdIds', () => {
    expect(repository.enumeration.get(entityName1).metaEdId).toBe(metaEdId1);
    expect(repository.enumeration.get(entityName2).metaEdId).toBe(metaEdId2);
  });
});

describe('when building enumeration with missing enumeration name', () => {
  const repository: Repository = repositoryFactory();
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = '';
  const documentation: string = 'Doc';
  const itemShortDescription: string = 'ItemShortDescription';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName)
      .withDocumentation(documentation)
      .withEnumerationItem(itemShortDescription)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should have missing id error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building enumeration with lowercase enumeration name', () => {
  const repository: Repository = repositoryFactory();
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'entityName';
  const documentation: string = 'Doc';
  const itemShortDescription: string = 'ItemShortDescription';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName)
      .withDocumentation(documentation)
      .withEnumerationItem(itemShortDescription)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should have extraneous input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building enumeration with missing documentation', () => {
  const repository: Repository = repositoryFactory();
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const itemShortDescription: string = 'ItemShortDescription';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName)
      .withEnumerationItem(itemShortDescription)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should have mismatched input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building enumeration with missing enumeration item', () => {
  const repository: Repository = repositoryFactory();
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const documentation: string = 'Doc';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName)
      .withDocumentation(documentation)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should have mismatched input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building enumeration with empty enumeration item description', () => {
  const repository: Repository = repositoryFactory();
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'entityName';
  const documentation: string = 'Doc';
  const itemShortDescription: string = '';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName)
      .withDocumentation(documentation)
      .withEnumerationItem(itemShortDescription)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should have mismatched input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});

describe('when building enumeration with invalid trailing text', () => {
  const repository: Repository = repositoryFactory();
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespace: string = 'namespace';
  const projectExtension: string = 'ProjectExtension';

  const entityName: string = 'EntityName';
  const documentation: string = 'Doc';
  const itemShortDescription: string = 'ItemShortDescription';
  const trailingText: string = '\r\nTrailingText';

  beforeAll(() => {
    const builder = new EnumerationBuilder(repository);

    textBuilder
      .withBeginNamespace(namespace, projectExtension)
      .withStartEnumeration(entityName)
      .withDocumentation(documentation)
      .withEnumerationItem(itemShortDescription)
      .withTrailingText(trailingText)
      .withEndEnumeration()
      .withEndNamespace()
      .sendToListener(builder);
  });

  it('should have extraneous input error', () => {
    expect(textBuilder.errorMessages).toMatchSnapshot();
  });
});
