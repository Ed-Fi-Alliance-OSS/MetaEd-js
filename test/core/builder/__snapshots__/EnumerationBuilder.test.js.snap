// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when building enumeration with empty enumeration item description should have mismatched input error 1`] = `
Array [
  "extraneous input 'entity' expecting ID, column: 14, line: 2, token: entity",
]
`;

exports[`when building enumeration with invalid trailing text should have extraneous input error 1`] = `
Array [
  "extraneous input 'TrailingText' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain'}, column: 0, line: 6, token: TrailingText",
]
`;

exports[`when building enumeration with lowercase enumeration name should have extraneous input error 1`] = `
Array [
  "extraneous input 'entity' expecting ID, column: 14, line: 2, token: entity",
]
`;

exports[`when building enumeration with missing documentation should have mismatched input error 1`] = `
Array [
  "mismatched input 'item' expecting {'documentation', METAED_ID}, column: 4, line: 3, token: item",
]
`;

exports[`when building enumeration with missing enumeration item should have mismatched input error 1`] = `
Array [
  "mismatched input 'End Namespace' expecting 'item', column: 0, line: 5, token: End Namespace",
]
`;

exports[`when building enumeration with missing enumeration name should have missing id error 1`] = `
Array [
  "missing ID at 'documentation', column: 4, line: 3, token: documentation",
]
`;
